File: src/web/Utils/CookieEvents.cs
using System.Security.Claims;
using System.Threading.Tasks;
using core.fs.Accounts;
using Microsoft.AspNetCore.Authentication;
using Microsoft.AspNetCore.Authentication.Cookies;
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.Logging;

namespace web.Utils
{
    public class CookieEvents : CookieAuthenticationEvents
    {
        private readonly ILogger<CookieEvents> _logger;
        private readonly Handler _service;

        public CookieEvents(
            ILogger<CookieEvents> logger,
            Handler service)
        {
            _logger = logger;
            _service = service;
        }

        public override Task RedirectToLogin(RedirectContext<CookieAuthenticationOptions> ctx)
        {
            if (ctx.Request.Path.StartsWithSegments("/api"))
            {
                ctx.Response.StatusCode = StatusCodes.Status401Unauthorized;
            }

            return base.RedirectToLogin(ctx);
        }
        
        public override async Task SigningIn(CookieSigningInContext context)
        {
            var query = new LookupByEmail(context.Principal.Email());

            var response = await _service.Handle(query);

            if (response.IsOk == false)
            {
                _logger.LogCritical($"Unable to look up user {query.Email} for sign in");
                throw new System.Exception("Failed to sign in via google");
            }
            
            if (context.Principal is not { Identity: ClaimsIdentity identity })
            {
                _logger.LogCritical("Claims principal is not a claims identity, it's a {principal}", context.Principal?.GetType().Name);
                throw new System.Exception("Failed to sign in via google");
            }

            identity.AddClaim(
                new Claim(IdentityExtensions.ID_CLAIM_NAME, response.ResultValue.Id.ToString())
            );
        }

        public override async Task ValidatePrincipal(CookieValidatePrincipalContext context)
        {
            var query = new LookupByEmail(context.Principal.Email());

            var id = await _service.Handle(query);
            if (id.IsOk == false)
            {
                _logger.LogCritical($"Failed to validate principal");
                context.RejectPrincipal();
                await context.HttpContext.SignOutAsync(CookieAuthenticationDefaults.AuthenticationScheme);
            }
        }
    }
}

File: src/web.fs/Utils/CookieEvents.fs
namespace web.Utils

open System.Security.Claims
open System.Threading.Tasks
open core.fs.Accounts
open Microsoft.AspNetCore.Authentication
open Microsoft.AspNetCore.Authentication.Cookies
open Microsoft.AspNetCore.Http
open Microsoft.Extensions.Logging

type CookieEvents(logger: ILogger<CookieEvents>, service: Handler) =
    inherit CookieAuthenticationEvents()
    
    override this.RedirectToLogin(ctx: RedirectContext<CookieAuthenticationOptions>) =
        if ctx.Request.Path.StartsWithSegments("/api") then
            ctx.Response.StatusCode <- StatusCodes.Status401Unauthorized
        
        base.RedirectToLogin(ctx)
    
    override this.SigningIn(context: CookieSigningInContext) = task {
        let query = LookupByEmail(context.Principal.Email())
        
        let! response = service.Handle(query)
        
        match response with
        | Error _ ->
            logger.LogCritical($"Unable to look up user {query.Email} for sign in")
            raise (System.Exception("Failed to sign in via google"))
        | Ok result ->
            match context.Principal with
            | null -> 
                logger.LogCritical("Claims principal is null")
                raise (System.Exception("Failed to sign in via google"))
            | principal ->
                match principal.Identity with
                | :? ClaimsIdentity as identity ->
                    identity.AddClaim(
                        Claim(ID_CLAIM_NAME, result.Id.ToString())
                    )
                | _ ->
                    logger.LogCritical($"Claims principal is not a claims identity, it's a {principal.GetType().Name}")
                    raise (System.Exception("Failed to sign in via google"))
    }
    
    override this.ValidatePrincipal(context: CookieValidatePrincipalContext) = task {
        let query = LookupByEmail(context.Principal.Email())
        
        let! id = service.Handle(query)
        match id with
        | Error _ ->
            logger.LogCritical("Failed to validate principal")
            context.RejectPrincipal()
            do! context.HttpContext.SignOutAsync(CookieAuthenticationDefaults.AuthenticationScheme)
        | Ok _ -> ()
    }
